app/services/github/client.rb

```diff
@@ -0,0 +1,103 @@
+module Github
+  # Github client (uses ocktokit.rb as a backend)
+  class Client
+    class << self
+      # adapted from https://api.rubyonrails.org/classes/Module.html#method-i-delegate_missing_to
+      def method_missing(method, *args, &block)
+        return super unless target.respond_to?(method, false)
+
+        # define for re-use
+        self.class.define_method(method) do |*new_args, &new_block|
+          request do
+            target.public_send(method, *new_args, &new_block)
+          end
+        end
+
+        # call the original method, this will only be called the first time
+        # as in subsequent calls, the newly defined method will prevail
+        request do
+          target.public_send(method, *args, &block)
+        end
+      end
+
+      # adapted from https://api.rubyonrails.org/classes/Module.html#method-i-delegate_missing_to
+      def respond_to_missing?(method, _include_all = false)
+        target.respond_to?(method, false) || super
+      end
+
+      private
+
+      def request
+        Honeycomb.add_field("app.name", "github.client")
+        yield
+      rescue Octokit::Error => e
+        record_error(e)
+        handle_error(e)
+      end
+
+      def record_error(exception)
+        class_name = exception.class.name.demodulize
+
+        Honeycomb.add_field("github.result", "error")
+        Honeycomb.add_field("github.error", class_name)
+        DatadogStatsClient.increment(
+          "github.errors",
+          tags: ["error:#{class_name}", "message:#{exception.message}"],
+        )
+      end
+
+      def handle_error(exception)
+        class_name = exception.class.name.demodulize
+
+        # raise specific error if known, generic one if unknown
+        error_class = "::Github::Errors::#{class_name}".safe_constantize
+        raise error_class, exception.message if error_class
+
+        error_class = if exception.class < Octokit::ClientError
+                        Github::Errors::ClientError
```

Comment:
`Errors` feels redundant here, I think these could be 
```
Github::ClientError
Github::ServerError
Github::Error
```

Reply:
true, but I followed the convention in `Search::`, not that it can't be changed, we just need to make a decision about this or ignore it until we make our minds up :D

Reply:
Maybe followup Refactor PR to remove redundant Error from everything 😄

Reply:
I just thought that there might be an advantage though to the namespacing: it's easier to do runtime reflection. Don't know if it's enough to justify it though ahhahaha